src/api.ts

```diff
@@ -195,69 +236,264 @@ function createApi({
+        })
+
+        if (data) api.putTileset(id, data)
+      }
+
+      if (row.upstreamUrl) {
+        fetchOnlineResource(row.upstreamUrl, row.etag).catch(() => {})
       }
+
+      return { ...JSON.parse(row.tilejson), id }
     },
 
     async getTile({ tilesetId, zoom, x, y }) {
-      const tilestore = context.tilestores.get(tilesetId)
-      if (!tilestore) {
-        throw new NotFoundError(tilesetId)
+      const quadKey = tileToQuadkey([x, y, zoom])
+
+      let tile:
+        | {
+            data: Buffer
+            etag?: string
+          }
+        | undefined = db
+        .prepare<{
+          tilesetId: string
+          quadKey: string
+        }>(
+          'SELECT TileData.data as data, Tileset.etag as etag FROM TileData ' +
+            'JOIN Tile ON TileData.tileHash = Tile.tileHash ' +
+            'JOIN Tileset ON Tile.tilesetId = Tileset.id ' +
+            'WHERE Tileset.id = :tilesetId AND Tile.quadKey = :quadKey'
+        )
+        .get({ tilesetId, quadKey })
+
+      async function fetchOnlineResource(): Promise<
+        | {
+            data: Buffer
+            etag?: string
+          }
+        | undefined
+      > {
+        const upstreamTileUrl = getUpstreamTileUrl({
+          tilesetId,
+          zoom,
+          x,
+          y,
+        })
+
+        // TODO: Need to check if we can make online requests too
+        if (upstreamTileUrl) {
+          const response = await upstreamRequestsManager.getUpstream<Buffer>({
+            url: upstreamTileUrl,
+            etag: tile?.etag,
+            responseType: 'buffer',
+          })
+
+          if (response) {
+            await api.putTile({
+              tilesetId,
+              zoom,
+              x,
+              y,
+              data: response.data,
+              etag: response.etag,
+            })
+
+            return { data: response.data, etag: response.etag }
+          }
+        }
+      }
+
+      if (tile) {
+        fetchOnlineResource().catch(() => {})
+      } else {
+        const response = await fetchOnlineResource()
+
+        // TODO: Should this throw if false?
+        if (response) {
+          tile = response
+        }
+      }
+
+      if (!tile) {
+        // TODO: Improve error handling here?
+        throw new NotFoundError(
+          `Tileset id = ${tilesetId}, [${zoom}, ${x}, ${y}]`
+        )
+      }
+
+      return {
+        data: tile.data,
+        // TODO: This never returns a Last-Modified header but seems like the endpoint would want it to if possible?
+        // Would require changing the return type of UpstreamRequestsManager.getUpstream
+        headers: { ...headers(tile.data), Etag: tile.etag },
       }
-      return tilestore.getTile(zoom, x, y)
+    },
+
+    async putTile({ tilesetId, zoom, x, y, data, etag }) {
+      const upstreamTileUrl = getUpstreamTileUrl({
+        tilesetId,
+        zoom,
+        x,
+        y,
+      })
+
+      const quadKey = tileToQuadkey([x, y, zoom])
```

Comment:
agreed – this is a function export from `@mapbox/tilebelt` but can wrap it to avoid potential issues if desired

Reply:
better yet, i'll just copy the implementation from tilebelt so that we can remove tilebelt as a dep – we're not using it for anything but this function

Reply:
Agreed - see https://macwright.com/2021/03/11/vendor-by-default.html